jvm-jersey-env

sahil-lakhwani opened the pull request:
Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

This PR adds a new JVM environment.
This JVM environment is based  on the Jersey framework, giving a bare minimum implementation of JAX-RS.

The user function is passed a [ContainerRequestContext](https://docs.oracle.com/javaee/7/api/javax/ws/rs/container/ContainerRequestContext.html) object and is expected to return a [Response](https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.html) object.

Example user function can be found under `examples/jvm-jersey`. It demonstrates both, a GET and a POST receiver.

Depends on https://github.com/fission/fission-java-libs/pull/2
The new fission-java-libs need to be published for this PR code to work.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fission/fission/1677)
<!-- Reviewable:end -->

vishal-biyani commented on environments/jvm-jersey/src/main/java/io/fission/JerseyServer.java:
```diff
@@ -0,0 +1,147 @@
+		String entryPoint = req.getFunctionName();
+		logger.log(Level.INFO, "Entrypoint class:" + entryPoint);
+		if (entryPoint == null) {
+			return Response.status(Response.Status.BAD_REQUEST).entity("Entrypoint class is missing in the function")
```
```suggestion
			return Response.status(Response.Status.BAD_REQUEST).entity("Entrypoint class is missing in the function JAR or the name is incorrect")
```

vishal-biyani commented on environments/jvm/pom.xml:
```diff
@@ -21,7 +21,7 @@
 		<dependency>
 			<groupId>io.fission</groupId>
-			<artifactId>fission-java-core</artifactId>
+			<artifactId>fission-java-spring</artifactId>
```
In order to support backward compatibility - we should let `fission-java-core` be as it is

vishal-biyani commented on environments/jvm/pom.xml:
```diff
@@ -43,7 +43,7 @@
 		<repository>
 			<id>fission-java-core</id>
-			<name>fission-java-core-snapshot</name>
+			<name>fission-java-spring-snapshot</name>
```
In order to support backward compatibility - we should let `fission-java-core` be as it is

sahil-lakhwani commented on environments/jvm-jersey/pom.xml:
```diff
@@ -0,0 +1,95 @@
+	<version>0.0.1-SNAPSHOT</version>
+
+	<properties>
+		<java.source.level>1.6</java.source.level>
+		<java.target.level>1.6</java.target.level>
```
Yes, for Java 9 and above, anything below 1.6 is not valid.

sahil-lakhwani commented on environments/jvm-jersey/pom.xml:
```diff
@@ -0,0 +1,95 @@
+			<artifactId>jersey-container-servlet-core</artifactId>
+			<version>2.0</version>
+		</dependency>
+		<dependency>
+			<groupId>org.glassfish.jersey.media</groupId>
+			<artifactId>jersey-media-json-jackson</artifactId>
+			<version>2.0</version>
```
Yes it is required.

sahil-lakhwani commented on environments/jvm-jersey/pom.xml:
```diff
@@ -0,0 +1,95 @@
+			<version>2.0</version>
+		</dependency>
+		<dependency>
+			<groupId>javax.xml</groupId>
+			<artifactId>jaxb-api</artifactId>
+			<version>2.1</version>
```
Yes.

vishal-biyani replied:
@sahil-lakhwani Do we have a test case for XML?

vishal-biyani commented:
Tested the XML example for this: https://github.com/fission/fission/pull/1541 and with this environment, it works great. I was able to get an entire POM.xml tested against Java env without any change: 
![image](https://user-images.githubusercontent.com/1167948/90638063-706efa80-e24a-11ea-800a-8be4abd2d996.png)

sahil-lakhwani commented on environments/jvm-jersey/src/main/java/io/fission/JerseyServer.java:
```diff
@@ -0,0 +1,147 @@
+		} catch (IllegalAccessException e) {
+			e.printStackTrace();
+			return Response.status(Response.Status.BAD_REQUEST)
+					.entity("Error creating a new instance of function class").build();
```
Yes, it will be.

sahil-lakhwani commented on environments/jvm-jersey/src/main/java/io/fission/JerseyServer.java:
```diff
@@ -0,0 +1,147 @@
+			Enumeration<JarEntry> e = jarFile.entries();
+			URL[] urls = { new URL("jar:file:" + file + "!/") };
+
+			// TODO Check if the classloading can be improved for ex. use something like:
+			// Thread.currentThread().setContextClassLoader(cl);
```
I think it is still relevant and should be addressed in #1674

sahil-lakhwani commented on environments/jvm-jersey/src/main/java/io/fission/JerseyServer.java:
```diff
@@ -0,0 +1,147 @@
+		return callUserFunction(request);
+	}
+
+	@Path("v2/specialize")
```
The class is annotated with a `/` so I think it is not necessary here.

sahil-lakhwani commented:
@vishal-biyani  Some changes after last review:
- Added Dockerfile for JVM 11 and related release commands in release.sh and release-build.sh
- Added new builder Dockerfiles under `environments/jvm-jersey/builder/ `
- Added new test case for the new jersey-env which also tests a POST request with XML body.

vishal-biyani commented:
@sahil-lakhwani Please resolve conflicts when you get a chance.

vishal-biyani commented:
Hey @sahil-lakhwani Can you please remove java-src-pkg.zip from commit?

sahil-lakhwani commented:
> Hey @sahil-lakhwani Can you please remove java-src-pkg.zip from commit?

Thanks for noticing, removed this.

vishal-biyani commented:
Thanks a ton, @sahil-lakhwani - this is a big improvement to the JVM environment and appreciate your efforts!
